Template Part Block: Add a block label

jeyip opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/master/docs/contributors/repository-management.md#pull-requests. -->

## Description
<!-- Please describe what you have changed or added -->
When the child of a template part block is selected, it can be unclear which template part we're modifying.

<img width="1015" alt="Screen Shot 2020-08-12 at 2 18 35 PM" src="https://user-images.githubusercontent.com/5414230/90070662-15b52c00-dca9-11ea-9cf5-bcd3d3c6c2d1.png">
<img width="1015" alt="Screen Shot 2020-08-12 at 2 18 42 PM" src="https://user-images.githubusercontent.com/5414230/90070667-16e65900-dca9-11ea-91ed-91de8ca2fd53.png">

This PR prototypes template part labels when a child block is selected (as discussed [here](https://github.com/WordPress/gutenberg/issues/22064#issuecomment-673258312))

### Mocks
<img width="483" alt="Screen Shot 2020-08-11 at 9 59 26 AM" src="https://user-images.githubusercontent.com/5414230/89926227-6c433d00-dbb9-11ea-9360-d7f409965d25.png">

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->
1. Set up the site editor locally
2. Navigate to the site editor page
3. Identify a template part block
4. Select a child of a template part block
5. Verify that there is a computed label matching the name of the current template part at the top left corner of the block

## Screenshots <!-- if applicable -->
### When selecting a child of a template part block
<img width="1133" alt="Screen Shot 2020-08-13 at 3 18 59 PM" src="https://user-images.githubusercontent.com/5414230/90192759-6f385c00-dd78-11ea-95d5-52031f712d04.png">

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

jeyip committed:
Add template part label

noahtallen commented on packages/block-library/src/template-part/edit/notice.js:
```diff
@@ -0,0 +1,23 @@
+ */
+import { useEntityProp } from '@wordpress/core-data';
+
+export default function TemplatePartNotice( { postId, slug } ) {
```
I feel like "notice" might be the wrong name for this -- I almost think "label" is better. But names are hard, so not sure :)

jeyip replied:
I totally agree. I was about to push a rebase now with some updates 🙂

jeyip commented on packages/block-library/src/template-part/edit/label.js:
```diff
@@ -0,0 +1,27 @@
+/**
```
**Note:** I was thinking about making this a HOC if we plan to use it on other blocks on the future.

jeyip replied:
**Note:**

We can definitely talk about refactoring this for reusability if we end up moving forward with template part labels

jeyip committed:
Update label border styling

Addison-Stavlo commented on packages/block-library/src/template-part/editor.scss:
```diff
@@ -77,6 +77,34 @@
+
+	.wp-block-template-part__label-text {
+		color: $black;
+		font-size: 10px;
```
this font size seems a bit small, it might be good to use `$default-font-size` (which I believe is 13px).

There may be other places in this added css where it might make sense to use defined values from the base_styles package (https://github.com/WordPress/gutenberg/tree/master/packages/base-styles)

jeyip replied:
That's a great callout. Thanks Addie! Making the changes now.

Addison-Stavlo commented on packages/block-library/src/template-part/editor.scss:
```diff
@@ -77,6 +77,34 @@
+	}
+
+	.wp-block-template-part__label-text {
+		color: $black;
```
Im not sure if we want to define the color here.  It will be already inherited from a higher level (which currently is black but may update in other circumstances like use of a dark theme? 🤔 )

Addison-Stavlo commented on packages/block-library/src/template-part/edit/label.js:
```diff
@@ -0,0 +1,27 @@
+import { Icon } from '@wordpress/components';
+
+export default function TemplatePartLabel( { postId, slug } ) {
+	const [ title ] = useEntityProp(
+		'postType',
+		'wp_template_part',
+		'title',
+		postId
+	);
```
I don't have a strong opinion on this, but figured I would point it out.  It seems we are also using this `useEntityProp` for title in the name-panel component.  Maybe it would make sense to move it into index and pass it down to these two child components?  But maybe it doesn't really matter and is more simple this way.

jeyip replied:
I don't have a strong opinion either. If this becomes widely shared, I'd err on the side of removing implicit dependencies and moving `useEntityProp` out of the label component. The consumer can always pass the value through props.

Addison-Stavlo commented on packages/block-library/src/template-part/edit/index.js:
```diff
@@ -67,10 +75,15 @@ export default function TemplatePartEdit( {
-					postId={ postId }
-					hasInnerBlocks={ innerBlocks.length > 0 }
-				/>
+				<div style={ { position: 'relative' } }>
```
Maybe we should apply this in editor.scss instead of inline?

Addison-Stavlo reviewed (commented):
I am liking this, its working well! 🎉 

Some thoughts on other additions (probably future PRs if design likes this enough as is for now):
* A hover state that triggers a pop-up/tooltip next to the label that can have text describing how it behaves differently from other blocks.
* Refactoring this out of template part so we can use it for other blocks with their own custom descriptions.

jeyip committed:
Use default font sizes

jeyip committed:
Use predefined border width

jeyip committed:
Remove inline position relative styling

jeyip committed:
Remove explicit color style for label text

dubielzyk commented:
This is looking nice! Awesome job. Let's tweak the label slightly to make it same as the mocks (https://d.pr/i/DX7mKm).
- Font-size should be 10px (currently it's 13px)
- Font-family should be system font (currently it's using Fira Sans)
- Font-weight should be 500 (not 400)
- Letter-spacing should be 0.25px
- The height of the label should be 22px in total (currently it's sitting at 21px. Sorry 😂)
- The label border should match the selection border, so that's already good. Ignore that in the mock.

We're now left with an interesting issue (which is why I didn't wanna go to the label immediately): The label we added is another label similar to the [selected block label](https://d.pr/i/xCazSA). So we effectively have two labels for the same thing.

ntsekouras commented on packages/block-library/src/template-part/edit/index.js:
```diff
@@ -67,10 +75,15 @@ export default function TemplatePartEdit( {
-					postId={ postId }
-					hasInnerBlocks={ innerBlocks.length > 0 }
-				/>
+				<div className=".wp-block-template-part__container">
```
```suggestion
				<div className="wp-block-template-part__container">
```

jeyip replied:
Thank you!

Addison-Stavlo commented:
>Font-size should be 10px (currently it's 13px)

@dubielzyk - Im curious about this, why are we going with 10px?  He had it at 10px previously and it seemed incredibly small and inconsistent with the rest of the editor.  I had recommended the change to `$default_font_size` (which is the 13px) at some point last week.  Looking at the [font sizes used in this repo](https://github.com/WordPress/gutenberg/blob/master/packages/base-styles/_variables.scss), `$editor-font-size` is 16px and the smallest font size we seem to have defined for use is the mentioned 13px.  Is there a reason why we need to go smaller than all the font sizes defined and used across the repo?

> Font-family should be system font (currently it's using Fira Sans)

`$default-font` should work for this (it defines system font as 1st priority):
https://github.com/WordPress/gutenberg/blob/52cf528997f57a00a7727125e5c932644af58cc6/packages/base-styles/_variables.scss#L7

shaunandrews commented:
<img width="404" alt="image" src="https://user-images.githubusercontent.com/191598/90416666-c42de800-e080-11ea-9003-a588e00197e2.png">

The block toolbar of the first child appears on top of the new label.

--

Similarly, in Select mode the new label competes with the existing button:

<img width="265" alt="image" src="https://user-images.githubusercontent.com/191598/90417395-a319c700-e081-11ea-9de7-bb3a56efc14c.png">

--

<img width="288" alt="image" src="https://user-images.githubusercontent.com/191598/90416749-de67c600-e080-11ea-9da9-1eb470f5d6b2.png">

The new label seems to be picking up some theme styles. The screenshot above is with Twenty Nineteen active.

Also, the icon here doesn't really correlate to the Template Part block icon (the LEGO):

<img width="450" alt="image" src="https://user-images.githubusercontent.com/191598/90417174-664dd000-e081-11ea-8ce3-aac6ccff3dcc.png">

shaunandrews commented:
> Font-size should be 10px (currently it's 13px)
> Font-family should be system font (currently it's using Fira Sans)
> Font-weight should be 500 (not 400)
> Letter-spacing should be 0.25px

I wonder if there is an existing style we can use; 10px seems a little small, and I'm not sure we use letter-spacing elsewhere.

> The height of the label should be 22px

Where does this number come from? I struggle myself with the general "grid" in Gutenberg and 22px isn't a value I've come across.

shaunandrews commented on packages/block-library/src/template-part/editor.scss:
```diff
@@ -77,6 +80,33 @@
+
+	.wp-block-template-part__label-text {
+		font-size: $default-font-size;
+		text-transform: uppercase;
```
I don't think we need to make this all uppercase; We don't do that elsewhere (do we?), and uppercase words like this can be more difficult to read (you _lose_ the letter shapes).

jeyip replied:
> uppercase words like this can be more difficult to read (you lose the letter shapes)

Looking at your examples with Twenty Nineteen, `UNTITLED TEMPLATE PART` _does_ seem difficult to read.

> We don't do that elsewhere (do we?)

This is a very good point. 👍

jeyip replied:
Pushed changes to switch the label to title case

jeyip committed:
Fix template part container classname

Co-authored-by: Nik Tsekouras <ntsekouras@outlook.com>

MichaelArestad commented:
I just tested this out locally. Overall, the concept works for me. I don't think it's ready just yet as I think some iteration is needed around where the label appears in a variety of situations:

1. As Shaun pointed out above, the label can be mostly covered in Select mode. Should the label move to another corner in that situation? Should it remain visible at all in select mode?
2. When template parts are nested, the labels overlap like this:
<img width="225" alt="image" src="https://user-images.githubusercontent.com/1123119/90423083-18cd6500-e079-11ea-911f-4a743f63689e.png">
What should be the rules around these labels in situations like this? 

Is the icon supposed to be a reusable block icon for template parts?

jeyip commented:
> We're now left with an interesting issue (which is why I didn't wanna go to the label immediately): The label we added is another label similar to the selected block label. So we effectively have two labels for the same thing.

I see. Which brings us back to the question of why we have two modes in the first place. 😛

> As Shaun pointed out above, the label can be mostly covered in Select mode. Should the label move to another corner in that situation? Should it remain visible at all in select mode?

Great point! I'm removing it entirely in select mode (for the time being), but this is definitely an ongoing discussion. It seems like the selection block labels are genericized with "Group", "Link", "Site Title", so it doesn't make sense to me to display custom names at all.

> Also, the icon here doesn't really correlate to the Template Part block icon (the LEGO):

The icon is a placeholder. @shaunandrews @MichaelArestad My mistake for not clarifying in the description of the PR, but these code changes aren't meant to be the finalized version of what we implement. We're in the process of deciding whether or not these labels are best for improving the template part editing user experience.

More context can be found in this issue https://github.com/WordPress/gutenberg/issues/22064

jeyip committed:
Disable labels in navigation mode

noahtallen commented:
> When template parts are nested, the labels overlap like this:

Should we only show the label of the "current active" template part, not the grandparent template part? This makes sense to me, because when you edit the child template part, you are only saving changes to the child. You don't actually modify the grandparent template part from a technical POV.

> It seems like the selection block labels are genericized with "Group", "Link", "Site Title", so it doesn't make sense to me to display custom names at all.

I think we've made some changes to this in #23847, where the label in select mode uses the dynamic label of the block (which is something like `Template Part: Header` now)

> Also, the icon here doesn't really correlate to the Template Part block icon

One of the ideas with the icon here is to communicate "reusability" and that the template part is a bit unique in that regard. IMO, it might be possible to find an icon which could to indicate reusability across a wide variety of blocks. Example: how does the user know that their site title block will actually have a very wide affect across the entirety of their site? Perhaps an icon could help indicate that the site title block is special in that way.

Really happy to see some progress here!

dubielzyk commented:
Wonderful comments @shaunandrews and @MichaelArestad . Thanks everyone. I'll work on improvements, but below are a few label alternatives I've looked at. Some definitely have the same issues so ignore those 😂
![image](https://user-images.githubusercontent.com/2896062/90470372-9c3f9280-e11b-11ea-8265-2dc6f8432df6.png)

That being said, while the idea of a Label might feel like it belongs more in the _Select mode_ I really enjoy playing with this PR. The label only shows when you're editing a template part and the label is still pretty subtle. I feel like it adds clarity to what you're actually editing.
![image](https://user-images.githubusercontent.com/2896062/90470417-c5602300-e11b-11ea-8935-25752a4bed1a.png)

I'm working on some further communication and messaging ideas and one that comes to mind is putting some info in the Block inspector. I'll create a separate issue for this 😃

MichaelArestad commented:
> That being said, while the idea of a Label might feel like it belongs more in the Select mode I really enjoy playing with this PR. 

To clarify, I don't think the label is necessary in select mode. Just edit mode.

jeyip commented:
> Should we only show the label of the "current active" template part, not the grandparent template part? This makes sense to me, because when you edit the child template part, you are only saving changes to the child. You don't actually modify the grandparent template part from a technical POV.

I'm curious what @shaunandrews and @MichaelArestad think, but this makes a lot of sense to me @noahtallen. I'll try to explore this idea.

MichaelArestad commented:
> Should we only show the label of the "current active" template part, not the grandparent template part?

@jeyip Showing the label of only the current active template part makes sense to me and further clarifies to the user what they are editing.

dubielzyk commented:
>Should we only show the label of the "current active" template part, not the grandparent template part?

I'm not sure I understand the difference. My original intentions was to only show the label in Edit mode when you're editing a child within a template part. Then we show the label on the template part that's being affected by that child edit. If that makes sense?

jeyip commented:
>I'm not sure I understand the difference. My original intentions was to only show the label in Edit mode when you're editing a child within a template part. Then we show the label on the template part that's being affected by that child edit. If that makes sense?

That totally does! What Michael was pointing out in his screenshot was a bit of an edge case. He was visualizing the situation where a user might create a child block _within_ a template part block, which would all be nested within _another_ template part block.

Child Block > Parent Block (Template Part) > Grandparent Block (Another Template Part)

<img width="225" alt="image" src="https://user-images.githubusercontent.com/1123119/90423083-18cd6500-e079-11ea-911f-4a743f63689e.png">

jeyip committed:
Only display labels on parent, and not ancestor, template parts

In a scenario where a template part with content is nested within another template part, only the direct parent's label should be displayed because only the direct parent has it's data modified from a technical point of view.

jeyip commented on packages/block-editor/src/store/selectors.js:
```diff
@@ -460,6 +460,21 @@ export const getBlockParents = createSelector(
+ *
+ * @return {string} ClientID of the parent block.
+ */
+export const getBlockParent = createSelector(
```
**Note:**

After combing through `core/block-editor` selectors, I couldn't find one that would only provide information about the parent, and not the rest of the ancestors. This can be changed if someone has better suggestions.

noahtallen replied:
this seems like a legitimate change to me if the behavior doesn't exist yet 👍 

I'm curious how the "select parent block" functionality works without this, though??

jeyip replied:
> I'm curious how the "select parent block" functionality works without this, though??

Great question! I'll look into it. That might give us insight into a selector we can use 🙂

jeyip replied:
The only reference I could find to select parent block was in [entity-record-item.js L42](https://github.com/WordPress/gutenberg/blob/master/packages/editor/src/components/entities-saved-states/entity-record-item.js#L42). It looks like they select the parent block by deriving information about the direct parent from [lines 16 - 29](https://github.com/WordPress/gutenberg/blob/master/packages/editor/src/components/entities-saved-states/entity-record-item.js#L16-L29).

This is an option if we don't want to include a new selector to the block-editor store.

jeyip commented on packages/block-library/src/template-part/edit/index.js:
```diff
@@ -30,20 +30,21 @@ export default function TemplatePartEdit( {
 			const {
 				getBlocks,
-				hasSelectedInnerBlock: getHasSelectedInnerBlock,
+				getSelectedBlockClientId,
```
**Note:**

This behavior still needs to support multi select

noahtallen commented on packages/block-editor/src/store/selectors.js:
```diff
@@ -460,6 +460,21 @@ export const getBlockParents = createSelector(
+	( state, clientId ) => {
+		return state.blocks.parents[ clientId ];
+	},
+	( state ) => [ state.blocks.parents ]
```
Not totally sure, maybe we should change this to `( state, clientId ) => [ state.blocks.parents[ clientID ] ]` so that it only refreshes on changes to its own parent, not any parent

dubielzyk commented:
>Noah proposed that, in such a situation, we explore only displaying the label for the direct parent.

Yeah, I like that solution :)

jeyip committed:
Title case the label

jeyip committed:
Modify temporary template part label icon

jeyip committed:
Display the label, even when children are multi selected

jeyip commented:
This is ready for another round of review. I've addressed most of the suggestions except for one issue. 

<img width="404" alt="image" src="https://user-images.githubusercontent.com/191598/90416666-c42de800-e080-11ea-9003-a588e00197e2.png">

`The block toolbar of the first child appears on top of the new label.`

I'm still mulling over how we can approach this 🤔

dubielzyk commented:
Is there a way we can detect the collision? One of my initial ideas had the label _below_ the template part. That would mean it's unlikely to be covered by the Gutenberg toolbar. Can we try that?

jeyip commented:
> Is there a way we can detect the collision? One of my initial explorations had the label below the template part.

There are definitely ways to make it work (some not prettier than others 😛). The real win would be if there's some sort of shared component for that, but I'm uncertain if it exists. @ockham @noahtallen @vindl any ideas?

In the meantime, I can go digging!

noahtallen commented:
I'm not aware of any built in things to detect that scenario

jeyip committed:
Render label at bottom when colliding with block toolbar

jeyip reviewed (commented):
I couldn't find any existing behavior in Gutenberg that detected elements overlapping. 

I implemented [a potential solution](https://github.com/WordPress/gutenberg/pull/24557/commits/791e11058aa87c1e49d5c896fbae701f8d614b77#diff-a2f080ab05e489bf77524d2fb88e8a6eR14-R42), but I think the biggest drawback is that the code relies on a specific classname to reference the toolbar element [here](https://github.com/WordPress/gutenberg/pull/24557/commits/791e11058aa87c1e49d5c896fbae701f8d614b77#diff-a2f080ab05e489bf77524d2fb88e8a6eR56-R58). It feels very brittle. I'll continue refactoring throughout the day.

With that being said, what do you think @noahtallen and @dubielzyk ? It seems like the label now reliably detects overlap with the toolbar.

![overlap detection](https://user-images.githubusercontent.com/5414230/90923317-534d3f80-e3a2-11ea-9b1b-692fd59f0a96.gif)

jeyip committed:
Only detect collision between label content and block toolbar

jeyip committed:
Refactor label overlap detection

We want to compare the template part label and block toolbar for overlap. Previously, we did so by selecting the block toolbar by classname. In some cases, however, the selector would be executed before the DOM was updated, returning an undefined reference.

We now handle selection of the block toolbar element differently. The block toolbar id is updated on every new block selection. Because of this, we, instead, save the programatically generated block toolbar id in the redux store. We then wait until the id is updated before reevaluating block toolbar overlap with the template part label.

noahtallen commented:
<img width="745" alt="Screen Shot 2020-08-24 at 3 23 00 PM" src="https://user-images.githubusercontent.com/6265975/91102291-b991c680-e61d-11ea-907d-539cb0fe8ca9.png">

Nice, that overlap detection is working very smoothly now! I think it will be useful to get design feedback on this (I'm a bit unsure about whether or not we should be moving it around quite a bit while the user is editing..)

shaunandrews commented:
The more I think about this, I think adding more "floating UI" into the canvas is likely the wrong approach. I think exposing the current template part in the top bar next to the document title ( `Homepage  •  Footer` ) might be better overall: #20877

jeyip commented:
> The more I think about this, I think adding more "floating UI" into the canvas is likely the wrong approach. I think exposing the current template part in the top bar next to the document title ( Homepage •  Footer ) might be better overall: #20877

@shaunandrews thanks for the input! I like the idea, particularly because the topbar is stickied and _always_ visible, which isn't true for the current mocks for template part labels.

I have some questions though.

1. It seems like, with the visual distance of the topbar to template part blocks, there might be more friction in associating the topbar label with the template we're editing. Could you elaborate on why "adding more "floating UI" into the canvas is likely the wrong approach"?
2. What would the display for nested template parts look like? Would it be something like (Homepage • Footer • FooterSection)?

@dubielzyk I'm curious what your thoughts are as well.

shaunandrews commented:
> Could you elaborate on why "adding more "floating UI" into the canvas is likely the wrong approach"?

Any UI we add into the canvas area is likely to affect the display of the document; In this case, the label would obscure other elements. Along the same lines, as the canvas starts to display a more realistic rendering of the front-end of a site, these sorts of floating UI will have to compete with a site's visual design.

I always take pause when I see more UI being introduce into the canvas itself.

dubielzyk commented:
To me adding it to the top bar seems like an unnatural place to show it, but it seems like adding a label to the canvas is going against a lot of principles and isn't getting overall agreement. I'm fine with moving forward with that solution. I'm working on the Document Settings (#20877) so I can have a deeper look at that solution.

Thanks for the feedback and thoughts @shaunandrews and @jeyip :)

jameskoster commented:
> The more I think about this, I think adding more "floating UI" into the canvas is likely the wrong approach.

+1 on this. I might even go one step further, and question why we're adding this at all. There are already two locations where the template part is exposed to users; the breadcrumb, and the block navigator. Do we need another one? If so, what problem is it solving?

To be clear, I totally appreciate why users need to understand the significance of editing template parts. But it seems to me that the important consequence of that behaviour is the impact it will have on the rest of the site when changes are saved. I'm not sure that displaying a label on the currently focused template part really conveys that, assuming that the average user likely doesn't know what a "template part" is.

noahtallen commented:
Part of the problem is that users simply don't use the block navigator or block breadcrumbs. Most folks I've seen use template parts _(including myself)_ have no clue what they are working with in the editor even though these tools exist! Maybe part of the solution is making those easier to access, but how? The breadcrumb is as obvious as it can be at the bottom, and the block navigator is not visible while you are editing anyways, so it isn't really a tool for "reorienting myself immediately."

Beyond this, I worry that our focus on removing stuff from the canvas is leading to a UX which is harder to understand and reason about. I think layout building requires tools, organization, and an understanding of how stuff works on your site. Part of that would include outlines and tools in-line while you are working with different site-level documents.

This is very different from post writing, which is all about removing distractions to focus on writing text. These are entirely different use cases, and other applications for those use cases reflect that. (e.g. MS word, google docs have little inline UI, whereas page builders and site builders often do have outlines and text tips to keep you oriented.)

In my mind, the site editor is less about writing content, and much more about designing a great site!

> In this case, the label would obscure other elements

Isn't this ok given that it would only show up while specifically editing the template part?

>  the important consequence of that behaviour is the impact it will have on the rest of the site when changes are saved

I think this is very true. But I don't think this PR specifically solves that problem. (Though I totally want to dig into solving that problem as well since it affects so many different blocks!) 

The problem we have currently is that _even with_ tools like the block navigator, it takes a lot of effort to determine exactly which blocks on the canvas are children of the template part block. Currently, without a border/label while editing, you would have to click on every block until you find the location where it switches to a different block being the parent. That's a pretty rough UX. I would even say that if you don't understand what a template part is, you wouldn't even know to try that approach! (And it would certainly not be familiar to users who don't use the block navigator or breadcrumb very often.)

The border solves part of that: while editing the template part, you know exactly where it begins and ends. I would suggest that the label solves the other part: exactly "what" does this border indicate? I think this is intuitive if you know what a template part is, and even if you don't, gives you enough information to work with it.

> I think exposing the current template part in the top bar next to the document title ( Homepage •  Footer ) might be better overall

I think the big problem with this is that it relies on users to know to go looking at the header to find the information, similar to the breadcrumb or block navigator. 

I want to mention again that we are very open to more designs and explorations around this! The thing is: the experience isn't great right now, and I think _something_ (like this PR along with the border) is better than nothing. I should also mention that I'm thinking about these problems from the standpoint of a beginner to WordPress (and even site building in general), so I'm very much in favor of being obvious about what's happening.

vindl commented:
> I might even go one step further, and question why we're adding this at all. There are already two locations where the template part is exposed to users; the breadcrumb, and the block navigator. Do we need another one? If so, what problem is it solving?

Based on some previous discussion I'm not sure if the breadcrumb is going to stick around in the long run, but even if it does it doesn't help much from my experience. It's not easy to tell at a glance if you are editing something in the header, especially when there is a lot of nesting and that line gets too long.

I think the problem with the block navigator, aside from discoverability that @noahtallen already mentioned, is that I'd have to toggle it back on whenever I want to make sure that I'm actually adding a block to the template part, and not to some other area. This will be cumbersome when adding multiple blocks.

dubielzyk commented:
Do you think something like a spotlight-like UI would work better, @jameskoster @shaunandrews ? Or is it the notion of adding any UI to the canvas _at all_?
![image](https://user-images.githubusercontent.com/2896062/91374714-186f5f80-e819-11ea-8edb-4680669cf180.png)

jameskoster commented:
> I think this is very true. But I don't think this PR specifically solves that problem. (Though I totally want to dig into solving that problem as well since it affects so many different blocks!)

You know what - you're right 😁 I think I'm getting too far ahead of myself, thinking about the holistic flow of editing template parts.

>  is that I'd have to toggle it back on whenever I want to make sure that I'm actually adding a block to the template part

Just to note, this may not always be the case: #22113

> Do you think something like a spotlight-like UI would work better

It's an interesting idea, but forcing spotlight mode feels a little heavy-handed to me. To avoid disrupting a users edit flow I wonder if the consequences should be revealed when they're finished - on save - using the multi-entity-saving patterns that are being worked on?

jeyip commented:
Not quite sure where we landed with these code changes @jameskoster @shaunandrews. Because of the experimental nature of the site editor (and the ability to easily revert changes), @noahtallen @Addison-Stavlo and I discussed the idea of merging this PR.

It would be a concrete change we could analyze and test in the context of other template part editing improvements, and if folks are still uncertain about it's utility in the future, we could roll back all the changes made here.

shaunandrews commented:
> Not quite sure where we landed with this PR

I'm not a fan, but I'm also not sure I have veto power :)

--

<img width="88" alt="image" src="https://user-images.githubusercontent.com/191598/91764301-0b4dca00-eba5-11ea-9ee4-de6a031c049a.png">

I'm still seeing some bleed-through of my theme's styles in the font, which is pretty small. This seems to be using a new icon I haven't seen before, and its pretty hard to recognize at this small size.

<img width="482" alt="image" src="https://user-images.githubusercontent.com/191598/91764475-5e278180-eba5-11ea-8ba2-2d96f99971ae.png">

In Top Toolbar mode, it can be difficult to understand what block is selected; I see the "Header" label and a subtle border around the Template Part block — but I have have the Site Title block selected.

noahtallen commented:
Do you think there is a better way to solve this problem?

jameskoster commented:
> In Top Toolbar mode, it can be difficult to understand what block is selected; I see the "Header" label and a subtle border around the Template Part block — but I have have the Site Title block selected.

This is a very good point. To the untrained eye it would be easy to assume that the header template part is selected there.

> Do you think there is a better way to solve this problem?

Perhaps we should concentrate efforts around the persistent block navigator seeing as that has the potential to solve this issue, and provide affordances for various other template editing activities as well?

davipontesblog commented:
> Do you think something like a spotlight-like UI would work better

I am a firm believer that a spotlight-like UI like this improves the understanding of template parts, and what exactly you are editing in the Site Editor. Not only it provides better focus (imagine a homepage layout from our current Gutenboarding designs in the site editor, with tens of blocks competing for attention, in their full colorful glory), but it allows users to learn their site structure as they edit.

jeyip commented:
> I see the "Header" label and a subtle border around the Template Part block — but I have the Site Title block selected.

> This is a very good point. To the untrained eye it would be easy to assume that the header template part is selected there.

This was deliberately introduced to provide context about which template part our blocks exist in.

| Template Part Block Selected | Template Part Child Block Selected |
| --- | --- |
| <img width="1015" alt="Screen Shot 2020-08-12 at 2 18 35 PM" src="https://user-images.githubusercontent.com/5414230/90070662-15b52c00-dca9-11ea-9cf5-bcd3d3c6c2d1.png"> | <img width="1015" alt="Screen Shot 2020-08-12 at 2 18 42 PM" src="https://user-images.githubusercontent.com/5414230/90070667-16e65900-dca9-11ea-91ed-91de8ca2fd53.png"> |

> Perhaps we should concentrate efforts around the persistent block navigator seeing as that has the potential to solve this issue, and provide affordances for various other template editing activities as well?

This is an interesting idea. For those with less context, here's a [link to the persistent block navigator issue](https://github.com/WordPress/gutenberg/issues/23939). I'm curious about a few things:

1. It seems like finding the persistent block navigator would be less obvious to the user, it might be less intuitive than labels (because of visual proximity to what the user is editing), and it would take up more visual real estate. I'm sure there are benefits, and I was hoping that @jameskoster @shaunandrews could elaborate.
2. What are the other affordances for template editing activities you had in mind @jameskoster ?

Addison-Stavlo commented:
> Perhaps we should concentrate efforts around the persistent block navigator seeing as that has the potential to solve this issue, and provide affordances for various other template editing activities as well?

From what I can tell this does aim to solve the same sort of issue, but from a different use case and perspective.  That block navigator will add some great improvements, but I don't see it making the need for on-page visual indicators to go away entirely.

> I am a firm believer that a spotlight-like UI like this improves the understanding of template parts, and what exactly you are editing in the Site Editor. Not only it provides better focus (imagine a homepage layout from our current Gutenboarding designs in the site editor, with tens of blocks competing for attention, in their full colorful glory), but it allows users to learn their site structure as they edit.

1000% agree with this.  Casting shade and/or dimming opacity on elements outside of the template part when it is selected feels very fitting for conveying that it is a separate entity from its surroundings.

vindl commented:
> Perhaps we should concentrate efforts around the persistent block navigator seeing as that has the potential to solve this issue, and provide affordances for various other template editing activities as well?

@jameskoster How will this interact with proposed sidebar navigation in https://github.com/WordPress/gutenberg/issues/23939? 🤔

shaunandrews commented:
> This was deliberately introduced to provide context about which template part our blocks exist in.

I imagine this won't be an issue if we add the current template part name next to the template name in https://github.com/WordPress/gutenberg/issues/20877

noahtallen commented:
> I imagine this won't be an issue if we add the current template part name next to the template name in #20877

I mean, this is very similar to the block breadcrumb, and lots of people don't know about that even though it is visible all the time :) (But I do agree that it would be a nice improvement!)

jeyip commented:
> I mean, this is very similar to the block breadcrumb, and lots of people don't know about that even though it is visible all the time :) (But I do agree that it would be a nice improvement!)

The template part name in the topbar also reminds me of breadcrumbs. I think that it's an improvement as well, but I think Addie made a great observation.

> From what I can tell this does aim to solve the same sort of issue, but from a different use case and perspective. That block navigator will add some great improvements, but I don't see it making the need for on-page visual indicators to go away entirely.

Maybe both would be warranted? The on-page visual indicators seem like they could complement the template part name in the topbar.

jameskoster commented:
> It seems like finding the persistent block navigator would be less obvious to the user

With all the nesting involved, I don't think it would be unreasonable to consider initially auto-exposing the block navigator when _template_ editing.

> and it would take up more visual real estate.

For just learning what template part you're editing at a glance I agree it is overkill. But that is what the breadcrumb is for.

> What are the other affordances for template editing activities you had in mind @jameskoster ?

In particular; quickly and accurately selecting heavily nested blocks, dragging/dropping between those nested groups, and adding blocks in the correct location in one click. I'm sure there are others to be explored :)

MichaelArestad commented:
> I mean, this is very similar to the block breadcrumb, and lots of people don't know about that even though it is visible all the time :) (But I do agree that it would be a nice improvement!)

@noahtallen I wouldn't compare it to the success of breadcrumbs. The location of the breadcrumbs make them near invisible. Putting the name in the top bar would be a visible change that could reduce the need for this PR.

I also strongly feel a persistent block navigator will be key to helping folks navigate the interface. It's something I'm particularly looking forward to. It also will alleviate this issue.

kellychoffman commented:
It seems like we're at a bit of impasse. The original problem still stands:

> When the child of a template part block is selected, it can be unclear which template part we're modifying.

_But_ this particular solution introduces an additional problem, which in my mind blocks the merging of this PR. I think its worth pausing this particular PR and going back to the drawing board to explore more options for solving this problem.

dubielzyk commented:
Agreed. Looks like this isn't the right way to go.

In the original issue I've suggested (https://github.com/WordPress/gutenberg/issues/22064#issuecomment-685186905) we try out a spotlight like mode for template parts (no label or borders). Perhaps we can try that out and see what people think?

noahtallen commented:
> I think its worth pausing this particular PR and going back to the drawing board to explore more options for solving this problem.

I agree with this since we are now at an impasse :) I do still feel this is one of the best options if the only worry is helping folks understand template parts, but I'll certainly concede it causes some other problems along the way. I look forward to hearing more ideas about how to solve it!

jeyip commented:
Thanks for the input everyone! 🙏 

I appreciate the honest and open discussion. I'll go ahead and close this PR for now. If things change, we can always revisit this PR.
